src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedEntryPointSymbol.cs

```diff
@@ -372,8 +372,37 @@ internal AsyncForwardEntryPoint(CSharpCompilation compilation, NamedTypeSymbol c
                         type: userMain.ReturnType)
                 { WasCompilerGenerated = true };
 
-                // The diagnostics that would be produced here will already have been captured and returned.
-                var success = binder.GetAwaitableExpressionInfo(userMainInvocation, out _getAwaiterGetResultCall!, runtimeAsyncAwaitCall: out _, _userMainReturnTypeSyntax, BindingDiagnosticBag.Discarded);
+                // Try to use the new HandleAsyncEntryPoint API if it exists
+                var specialMember = userMain.ReturnType.IsGenericTaskType(compilation)
+                    ? SpecialMember.System_Runtime_CompilerServices_AsyncHelpers__HandleAsyncEntryPoint_Task_T
+                    : SpecialMember.System_Runtime_CompilerServices_AsyncHelpers__HandleAsyncEntryPoint_Task;
+
+                if (Binder.TryGetSpecialTypeMember(compilation, specialMember, _userMainReturnTypeSyntax, BindingDiagnosticBag.Discarded, out MethodSymbol handleAsyncEntryPointMethod, isOptional: true))
+                {
+                    // Use the new HandleAsyncEntryPoint API
+                    _userEntryPointCall = new BoundCall(
+                            syntax: _userMainReturnTypeSyntax,
+                            receiverOpt: null,
+                            initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown,
+                            method: handleAsyncEntryPointMethod,
+                            arguments: ImmutableArray.Create<BoundExpression>(userMainInvocation),
```

Comment:
>userMainInvocation [](http://example.com/codeflow?start=78&length=18)

Is there a guarantee that the argument type matches? Could. for example, user's main return a user defined `Task` type distinct from one used by the API we are about to call?